### PR TITLE
9.X - async/await

### PR DESCRIPTION
**Pull request by ghermeto**
## Pre-Submission Checklist

- [x] Ran the linter and tests via `make prepush`
- [x] Included comprehensive and convincing tests for changes

# Changes

> What does this PR do?

_This PR has two **commits**:_
1. drops support to node 8.x (EOL);
2. adds support for async/await functions on `pre`, `use` and route handlers;

**_Example of async handling:_**

```js
    const restify = require('restify');
    const server = restify.createServer();
    
    server.pre(async function (req, res) {
        await something();
        doSomethingElse();
    });

    server.get('/async', async function (req, res) {
        const body = await thatThing();
        res.send(body);
    });
```

Basically, the `Chain` call will check if the handler return is a `Promise` (by checking the existence of `.then`, as Falcor Router does it) and resolves by calling `next` or rejects by calling `next with error`.

.

### CODE REVIEW

**Comment by ghermeto** [comment]
cc/ @josephharrington

**Comment by ghermeto** [comment]
ops... forgot to add examples on the docs (and build the docs)...

**Comment by ghermeto** [comment]
doh! the 3rd commit with async/await docs made this commit huge... if you can go into each individual commit to add your comments, I appreciate  :)

**Comment by ghermeto** [comment]
cc /@cprussin

**Comment by DonutEspresso** [comment]
Nit: For future, would recommend separate, smaller PRs for easier reviewing/accounting. There's a lot going on here (style changes, docs, Node 8.x EOL, etc) and its a bit hard to parse it all at once. 

A few questions that might influence the direction of the PR:
- Do we plan on supporting regular functions that return promises? Depending on the answer may have further implications to discuss, it doesn't look like we make a distinction at the moment.
- Do we plan on supporting mixed mode middleware? e.g., some callback, some async/await. I think the answer is yes, but good to be explicit.  
- I think the rejection contract could benefit from a bit more clarity/discussion. I want to say we should only allow rejecting with Error objects, but we can't enforce that at server startup. What are some of the ecosystem best practices here? 
- Async/await is likely to introduce some race conditions. e.g., imagine a `res.write` call in a loop. It looks like [fastify has some clear rules](https://www.fastify.io/docs/latest/Routes/#async-await) around this to ensure things don't go off the rails. Anything that applies to us or that we can learn from?

**Comment by ghermeto** [comment]
> Nit: For future, would recommend separate, smaller PRs for easier reviewing/accounting. There's a lot going on here (style changes, docs, Node 8.x EOL, etc) and its a bit hard to parse it all at once.

Yes, I broke them into different commits, but you are right. After I added the documentation commit  I realized it should have waited. I won't send multiple commits next time. 

> A few questions that might influence the direction of the PR:
> 
> * Do we plan on supporting regular functions that return promises? Depending on the answer may have further implications to discuss, it doesn't look like we make a distinction at the moment.

I followed the ecosystem. I guess we could check `handler.contructor.name === 'AsyncFunction'`, but it is [not recommended](https://github.com/tc39/ecmascript-asyncawait/issues/78#issuecomment-164425925). For instance, fastify [only checks the constructor name](https://github.com/fastify/fastify/blob/bf18412f2ba838fc06e3162a13ebcf88b5a283be/fastify.js#L398-L408) to verify the arity of the handler when inserted, but later they just [do this same check](https://github.com/fastify/fastify/blob/bf18412f2ba838fc06e3162a13ebcf88b5a283be/lib/hooks.js#L160) for `then`. I wouldn't oppose check for the constructor & arity, but I believe is unnecessary, since the result **_must be treated_** like any Promise.

> * Do we plan on supporting mixed mode middleware? e.g., some callback, some async/await. I think the answer is yes, but good to be explicit.

yes, this allows mixed mode. Like the other frameworks that support both models.

> * I think the rejection contract could benefit from a bit more clarity/discussion. I want to say we should only allow rejecting with Error objects, but we can't enforce that at server startup. What are some of the ecosystem best practices here?

I used the [same approach as fastify](https://github.com/fastify/fastify/blob/bf18412f2ba838fc06e3162a13ebcf88b5a283be/lib/hooks.js#L169-L175). You are right, we can't enforce it and it deviates from the current behavior. I don't think changing the error handling behavior should be coupled with this PR. 

What I can do is check for if  `err instanceof Error` or `err instanceof VError` and, if not, I can force wrap it under an `AsyncHandlerRejection` (or `AsyncHandlerError`), but that will differ from the callback behavior that allows user to throw anything.

> * Async/await is likely to introduce some race conditions. e.g., imagine a `res.write` call in a setInterval loop. It looks like [fastify has some clear rules](https://www.fastify.io/docs/latest/Routes/#async-await) around this to ensure things don't go off the rails. Anything that applies to us or that we can learn from?

The biggest problem with fastify happens because they allow the user to both return the function with the body or use `reply.send`. It provides a nicer DevEx but comes with trade-offs. You can't do this in restify. You still need to call `res.send`, but the `setInterval` is an interesting case.

We should recommend using `async/await` syntax on a `promise` or `async/await` context. If you use any code that relies on callbacks and still want to use async handlers, you should convert these to promises outside the handler.  Or better yet, avoid using async handlers. I guess that is something to append to the documentation. Maybe on a follow-up PR to add more docs around it?

**Comment by ghermeto** [comment]
@DonutEspresso Started issue #1837 to discuss this further.

File: docs/_api/formatters.md [hunk 1]
```diff
@@ -50,7 +50,7 @@ Format a response for being sent over the wire
 
 Type: [Function][9]
 
-**Parameters**
+#### Parameters
```
Comment (hekike) [inline docs/_api/formatters.md:53]: Could you please generate docs as a separate commit outside of this PR?

File: lib/chain.js [hunk 2]
```diff
@@ -157,19 +157,38 @@ function call(handler, err, req, res, _next) {
...
+    }
+
+    function reject(error) {
+        if (!error) {
+            error = new errors.VError(
+                {
+                    name: 'AsyncHandlerRejection',
+                    info: { handler: handler._name }
+                },
+                `Async middleware ${handler._name} rejected without an error`
```
Comment (hekike) [inline lib/chain.js:171]: Please avoid dynamic input in messages.

File: lib/chain.js [hunk 3]
```diff
@@ -157,19 +157,38 @@ function call(handler, err, req, res, _next) {
...
+            );
+        }
+        next(error);
```
Comment (hekike) [inline lib/chain.js:174]: nit: add return to be consistent and highlight this is the end

Comment (hekike) [inline lib/chain.js:174]: I think it's expected with async-await but this will catch syntax errors etc. 
Maybe explain in the 9.x migration guide.

File: lib/chain.js [hunk 4]
```diff
@@ -157,19 +157,38 @@ function call(handler, err, req, res, _next) {
...
+    function resolve() {
+        next();
+    }
+
+    function reject(error) {
+        if (!error) {
+            error = new errors.VError(
+                {
+                    name: 'AsyncHandlerRejection',
+                    info: { handler: handler._name }
```
Comment (hekike) [inline lib/chain.js:169]: Please add `cause`

File: package.json [hunk 5]
```diff
@@ -90,7 +90,7 @@
     "report-latency": "./bin/report-latency"
   },
   "engines": {
-    "node": ">=8.3.0"
+    "node": ">=10.21.0"
```
Comment (hekike) [inline package.json:93]: Please add a migration guide like [6to7](https://github.com/restify/node-restify/blob/master/docs/guides/6to7guide.md) and highlight Node version bump.

Comment (hekike) [inline package.json:93]: I think we need to change Travis CI minimum node test version as well

File: lib/chain.js [hunk 6]
```diff
@@ -157,19 +157,38 @@ function call(handler, err, req, res, _next) {
...
 
+    function resolve() {
+        next();
+    }
+
+    function reject(error) {
+        if (!error) {
+            error = new errors.VError(
+                {
+                    name: 'AsyncHandlerRejection',
```
Comment (hekike) [inline lib/chain.js:168]: Please define new error types in: https://github.com/restify/node-restify/blob/master/lib/errorTypes.js

**Comment by mmarchini** [comment]
@ghermeto if you want some help I can break this PR in three and force-push only one commit here. It might be easier for folks to review.

File: lib/chain.js (hunk 4 above)
Comment (ghermeto) [reply lib/chain.js:169]: what should add here as a `cause`? This error is created whenever the user rejects an async function without passing it anything:

```js
server.get('/params', async (req, res) => {
  return Promise.reject();
});
```
Do you have suggestions for `cause`?

Comment (ghermeto) [reply lib/chain.js:169]: You know... just got a mistake in my code... I wasn't wrapping synthetic errors because I thought Restify would throw anything, but actually, if the value passed to `next()` is a string, it will try to route it, for instance:

```js
const restify = require('restify');
const server = restify.createServer({});

server.use(restify.plugins.queryParser({mapParams:true}));

server.get('/route', (req, res, next) => {
   next('getresult');
});

server.get('/string', (req, res, next) => {
   next('invalid');
});

server.get('/number', (req, res, next) => {
   next(1);
});

server.get('/error', (req, res, next) => {
   next(new Error(''));
});

server.get('/result', (req, res, next) => {
   res.send(req.query);
   next();
});

server.listen(8081, 'localhost', function () {
   console.info('server started');
});

```

The result of `curl http://localhost:8081/error?id=1` is:
```json
{"code":"Internal","message":"Error"}
```
The result of `curl http://localhost:8081/number?id=1` is:
```json
{"code":"Internal","message":"1"}
```
The result of `curl http://localhost:8081/string?id=1` is:
```json
{"code":"Internal","message":"RouteMissingError: Route by name doesn't exist"}
```
The result of `curl http://localhost:8081/route?id=1` is:
```json
{"id":"1"}
```

I will update the code to always wrap rejections and update #1837 .

File: lib/chain.js (hunk 3 above)
Comment (ghermeto) [reply lib/chain.js:174]: Should the migration guide be part of a separate PR, with the docs?

File: package.json (hunk 5 above)
Comment (ghermeto) [reply package.json:93]: Should the migration guide be a separate PR?

File: lib/chain.js (hunk 4 above)
Comment (cprussin) [reply lib/chain.js:169]: FWIW trying to route on strings is IMHO an awful barnacle of Restify, it's worth considering deprecating it--it's essentially a GOTO with all the pain that comes along.

Comment (ghermeto) [reply lib/chain.js:169]: wrapping all non-errors in `AsyncHandlerError` actually turned out pretty nice, because if nothing is specified, the cause is `undefined`.

Comment (ghermeto) [reply lib/chain.js:169]: and I just discovered `cause` must be an instance of `Error`, so I'm moving the `cause` to inside the `info`.

File: package.json (hunk 5 above)
Comment (hekike) [reply package.json:93]: yes, pleace

**Comment by ghermeto** [comment]
thank you @mmarchini. I force pushed removing the generated docs and added some changes Peter requested. 
I kept the `jsdoc` updates as it made sense to be on the same PR, now I'm missing the migration guide.

**Comment by DonutEspresso** [comment]
> I followed the ecosystem. I guess we could check `handler.contructor.name === 'AsyncFunction'`, but it is [not recommended](https://github.com/tc39/ecmascript-asyncawait/issues/78#issuecomment-164425925). 

Yeah, that makes sense. My concern is that this opens up a new footgun with promise usage. The following handler will just hang forever:

```
function(req, res) {
  return FuncThatWasSupposedToReturnPromiseButDidnt();
}
```

I haven't used promises enough to know what's the best way to handle this. Is it common to do runtime assertions on thenables? I'd much rather the server blew up than fail silently. One way to enforce this is by saying async functions should only have an arity of 2, and we don't pass next to those functions. 

> > * I think the rejection contract could benefit from a bit more clarity/discussion. I want to say we should only allow rejecting with Error objects, but we can't enforce that at server startup. What are some of the ecosystem best practices here?
> 
> I used the [same approach as fastify](https://github.com/fastify/fastify/blob/bf18412f2ba838fc06e3162a13ebcf88b5a283be/lib/hooks.js#L169-L175). You are right, we can't enforce it and it deviates from the current behavior. I don't think changing the error handling behavior should be coupled with this PR. Right now, an `async` error will be treated exactly like a callback error.
> 
> What I can do is check for if `err instanceof Error` or `err instanceof VError` and, if not, I can force wrap it under an `AsyncHandlerRejection` (or `AsyncHandlerError`), but that will differ from the callback behavior that allows user to throw anything.

Sorry! I wasn't very clear, I was suggesting that we formalize the next() contract, and the promise pathways that lead up to that. Essentially, there's two ways to cede control back to restify through next today, and this PR adds a variation for each of those contracts: 

* next()
  * With async functions, a promise that resolves no value
* next('fooRoute') 
  * With async functions, a promise that resolves 'fooRoute'
* next(new Error('boom'))
  * With async functions, a promise that rejection with an Error object

I think there's a good argument to be made that anything that is not one of those three should cause a runtime exception. But I feel less strongly than I do about at least formalizing and documenting it.

**Comment by DonutEspresso** [comment]
> What I can do is check for if `err instanceof Error` or `err instanceof VError` and, if not, I can force wrap it under an `AsyncHandlerRejection` (or `AsyncHandlerError`), but that will differ from the callback behavior that allows user to throw anything.

What do you mean by `callback behavior that allows user to throw anything`?

**Comment by DonutEspresso** [comment]
Oops, sorry, maybe I should have left my comments in #1837!

**Comment by ghermeto** [comment]
I will reply to your comments on #1837

File: docs/guides/8to9guide.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+## Introduction
+
+restify `9.x` comes with `async/await` support!
+
+## Breaking Changes
+
+### Drops support for Node.js `8.x`
+
+Restify requires Node.js version `>=10.21.0`.
```
Comment (daviande) [inline docs/guides/8to9guide.md:14]: Is there anything in particular for 10.21.0 instead of just 10.x?

Comment (ghermeto) [reply docs/guides/8to9guide.md:14]: I guess not... I will update it

File: docs/guides/8to9guide.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+```js
+server.use(async (req, res) => {
+  req.something = await doSomethingAsync();
+});
+```
+- `fn.length === 2` (arity 2);
+- `fn instanceof AsyncFunction`;
+- if the async function resolves, it calls `next()`;
+- any value resolved will be discarded;
```
[the remark this answers is not in the document]
Comment (ghermeto) [reply docs/guides/8to9guide.md:47]: do you think it would read better "any value returned by the async function?". I will make the change.

File: docs/guides/8to9guide.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+- if it rejects with an `Error` instance it calls `next(err)`;
+- if it rejects with anything else it wraps in a `AsyncHandlerError` and calls `next(err)`;
+
+#### Route handler API
+
+```js
+server.use(async (req, res) => {
+  req.something = await doSomethingAsync();
+});
+```
```
[the remark this answers is not in the document]
Comment (ghermeto) [reply docs/guides/8to9guide.md:57]: OMG...

File: docs/guides/8to9guide.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+- if the async function resolves without a value, it calls `next()`;
+- if the async function resolves with a string value, it calls `next(string)` (re-routes*);
+- if the async function resolves with a value other than string, it calls `next(any)`;
+- if it rejects with an `Error` instance it calls `next(err)`;
+- if it rejects with anything else it wraps in a `AsyncHandlerError` and calls `next(err)` (error-handing**);
+
+##### (*) Note about re-routing:
+The `8.x` API allows re-routing when calling `next()` with a string value. If the string matches a valid route,
+it will re-route to that handler. The same is valid for resolving a async function. Whatever value it returns,
+will be sent to `next()` and the behavior will be exactly the same.
```
[the remark this answers is not in the document]
Comment (ghermeto) [reply docs/guides/8to9guide.md]: rephrased

File: lib/errorTypes.js [hunk 11]
```diff
@@ -5,5 +5,6 @@ var errors = require('restify-errors');
 // This allows Restify to work with restify-errors v6+
 module.exports = {
     RequestCloseError: errors.makeConstructor('RequestCloseError'),
-    RouteMissingError: errors.makeConstructor('RouteMissingError')
+    RouteMissingError: errors.makeConstructor('RouteMissingError'),
+    AsyncHandlerError: errors.makeConstructor('AsyncHandlerError')
```
[the remark this answers is not in the document]
Comment (ghermeto) [reply lib/errorTypes.js]: renamed

File: docs/guides/8to9guide.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ghermeto) [reply docs/guides/8to9guide.md]: fixed

File: docs/guides/8to9guide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ghermeto) [reply docs/guides/8to9guide.md]: done

File: docs/guides/8to9guide.md (hunk 7 above)
Comment (ghermeto) [reply docs/guides/8to9guide.md]: done

**Comment by ghermeto** [comment]
I spoke with @hekike and he said I can merge this. 

Just want to make sure @mmarchini (who had comments) and others can take a look at it before I do.

cc/ @m5m1th @josephharrington

**Comment by ghermeto** [comment]
ok, since no new objection was raised, and all comments were addressed, I'm merging this code. We can always submit fixes before we make a new release.

**Comment by nikeee** [comment]
What happens if the handler calls `next()` besides being async? For example:

```js
server.get('/async', async function (req, res) {
    const body = await thatThing();
    res.send(body);
    next();
});
```

Is this valid?
